### ISSUE TITLE
Generic Math - API tweaks

### ISSUE THREAD

**Issue by tannergooding** [dotnet/runtime#71648]
The Generic Math API surface is largely finalized and in the "correct" shape. There are a couple potential tweaks, however, that we should decide on where the implementation has had to deviate from the reviewed surface and design decisions.

### IUnaryPlusOperators

We are currently exposing `IUnaryPlusOperators` as the following:
```csharp
public interface IUnaryPlusOperators<TSelf, TResult>
    where TSelf : IUnaryPlusOperators<TSelf, TResult>
{
    static abstract TResult operator +(TSelf value);
}
```

This correctly matches the signature that C#/.NET allows for "unary plus" in that the input and return type do not have to match. However, since there isn't a way to convert between a `TSelf` and a `TResult` we can't expose a DIM for this as originallly intended.

We should decide whether to keep this "as is" or modify the API surface to the following so a DIM can be exposed:
```csharp
public interface IUnaryPlusOperators<TSelf>
    where TSelf : IUnaryPlusOperators<TSelf>
{
    static abstract TSelf operator +(TSelf value) => value;
}
```

Such a change would effectively enforce that "unary plus" is used as an "identity" or other simple operation.

### IComparisonOperators and IEqualityOperators

We are currently exposing `IComparisonOperators` and `IEqualityOperators` as the following:

```csharp
public interface IComparisonOperators<TSelf, TOther>
    : IComparable,
      IComparable<TOther>,
      IEqualityOperators<TSelf, TOther>
    where TSelf : IComparisonOperators<TSelf, TOther>
{
    static abstract bool operator <(TSelf left, TOther right);
    static abstract bool operator <=(TSelf left, TOther right);
    static abstract bool operator >(TSelf left, TOther right);
    static abstract bool operator >=(TSelf left, TOther right);
}

public interface IEqualityOperators<TSelf, TOther> : IEquatable<TOther>
    where TSelf : IEqualityOperators<TSelf, TOther>
{
    static abstract bool operator ==(TSelf left, TOther right);
    static abstract bool operator !=(TSelf left, TOther right);
}
```

This matches the expected and recommend signature for such operators where they always return a `bool`. However, this doesn't match the signature that C#/.NET more generally allow and so may be limiting to certain scenarios or other languages.

We should decide whether to keep this "as is" or modify the API surface to the allow the return type to also be generic:
```csharp
public interface IComparisonOperators<TSelf, TOther, TResult>
    : IComparable,
      IComparable<TOther>,
      IEqualityOperators<TSelf, TOther>
    where TSelf : IComparisonOperators<TSelf, TOther>
{
    static abstract TResult operator <(TSelf left, TOther right);
    static abstract TResult operator <=(TSelf left, TOther right);
    static abstract TResult operator >(TSelf left, TOther right);
    static abstract TResult operator >=(TSelf left, TOther right);
}

public interface IEqualityOperators<TSelf, TOther> : IEquatable<TOther>
    where TSelf : IEqualityOperators<TSelf, TOther>
{
    static abstract TResult operator ==(TSelf left, TOther right);
    static abstract TResult operator !=(TSelf left, TOther right);
}
```

### ITrigonometricFunctions

The `ITrigonmetricFunctions` interface currently exposes the following APIs:
```csharp
public interface ITrigonometricFunctions<TSelf>
    where TSelf : ITrigonometricFunctions<TSelf>, INumberBase<TSelf>
{
    static abstract TSelf Acos(TSelf x);
    static abstract TSelf AcosPi(TSelf x);
    static abstract TSelf Asin(TSelf x);
    static abstract TSelf AsinPi(TSelf x);
    static abstract TSelf Atan(TSelf x);
    static abstract TSelf Atan2(TSelf y, TSelf x);
    static abstract TSelf Atan2Pi(TSelf y, TSelf x);
    static abstract TSelf AtanPi(TSelf x);
    static abstract TSelf Cos(TSelf x);
    static abstract TSelf CosPi(TSelf x);
    static abstract TSelf Sin(TSelf x);
    static abstract (TSelf Sin, TSelf Cos) SinCos(TSelf x);
    static abstract TSelf SinPi(TSelf x);
    static abstract TSelf Tan(TSelf x);
    static abstract TSelf TanPi(TSelf x);
}
```

This is generally good/correct, however there are two potential concerns.

The first is that `Atan2` and `Atan2Pi` don't quite "work" for `Complex`. In particular, the signature for a `Complex` number should be `TScalar Atan2(Complex<TScalar> value)` rather than `TSelf Atan2(TSelf y, TSelf x)`. This is because its the equivalent of `TScalar Atan2(complex.Imaginary, complex.Real)`. It's proposed we don't include `Atan2` on `ITrigonometricFunctions` and instead isolate it to `IFloatingPointIeee754`. It would not be exposed on `IFloatingPoint` since that includes types like `decimal`, a currency type, where such operations don't make sense since they can't represent `NaN` or `Infinity`.

The second is that `SinCos` is exposed as a perf optimization, but we didn't consider `SinCosPi`. We could optionally split this into `ITrigonometricFunctions` and `ITrigonometricPiFunctions` to help keep surface area smaller and easier to opt into.

### IPowerFunctions and IRootFunctions

For `IRootFunctions` we only expose `TSelf Root(TSelf x, int n)` while for `IPowerFunctions` we only expose `TSelf Pow(TSelf x, TSelf y)`. The former matches `IEEE 754` expectations while the latter only matches the defined `pow` function. IEEE 754 also defines `pown` and `powr`. The former requires that `y` be an integer, simplifying the overall implementation and considerations. The latter disallows negative `x` as it only considers `exp(y * log(x))`. Both of these functions are expected to be approved as part of the next version of the C standard (tentatively C23) and it may be worth defining and exposing `Pow(TSelf x, int n)` or at least considering and reserving an "alternative" name for such variants to avoid future versioning concerns. If we reserve a name, we should consider also qualifying `Root` (`rootn` in C) and `Compound` (`compoundn` in C).

### I*Functions and math constant access

We are currently exposing several interfaces in the form of:
```csharp
public interface I*Functions<TSelf>
        where TSelf : I*Functions<TSelf>, INumberBase<TSelf>
{
}
```

These signatures are generally good and correct. However, there is a potential short term versioning concern here. In particular, some of these such as `Trigonometric` or `Exponential` don't have direct access to mathematical constants like `E` or `Pi`. However, given the additional constraint to `INumberBase`, they can still create approximations of such constants using something like `CreateSaturating(double.Pi)` and may in the future be able to do a `if (T is IFloatingPointIeee754)` or similar check to gain access to `T.Pi`.

We should determine if this is acceptable or if there should be one additional interface exposed:
```csharp
public interface IMathConstants<TSelf>
    where TSelf : IMathConstants<TSelf>
{
    public static TSelf E { get; }

    public static TSelf Pi { get; }

    public static TSelf Tau { get; }
}
```

Its worth noting that many standard C libraries define more than just these constants and so this could be expanded in the future:

| Native Name | .NET Name | Description                                                             |
| ----------- | --------- | ----------------------------------------------------------------------- |
| M_E         | E         | The base of natural logarithms.                                         |
| M_LOG2E     |           | The logarithm to base 2 of M_E.                                         |
| M_LOG10E    |           | The logarithm to base 10 of M_E.                                        |
| M_LN2       |           | The natural logarithm of 2.                                             |
| M_LN10      |           | The natural logarithm of 10.                                            |
| M_PI        | Pi        | Pi, the ratio of a circle’s circumference to its diameter.              |
| M_PI_2      |           | Pi divided by two.                                                      |
| M_PI_4      |           | Pi divided by four.                                                     |
| M_1_PI      |           | The reciprocal of pi (1/pi)                                             |
| M_2_PI      |           | Two times the reciprocal of pi.                                         |
| M_2_SQRTPI  |           | Two times the reciprocal of the square root of pi.                      |
| M_SQRT2     |           | The square root of two.                                                 |
| M_SQRT1_2   |           | The reciprocal of the square root of two (also the square root of 1/2). |
|             | Tau       | Pi times two                                                            |

**Comment by hopperpl** [comment]
May I ask that we also add Tau (2 * Pi) to the constants? 

Tau is already available in Math and MathF, therefore this would not introduce a new constant "type".

**Comment by tannergooding** [comment]
`Tau` is already exposed as part of `IFloatingPointIeee754` and therefore `float.Tau`, `double.Tau`, and `Half.Tau` are all available.

**Comment by hopperpl** [comment]
@tannergooding  I think you just added a mistake when updating the table.

M_2_PI is 2/π (not 2π or Tau)

**Comment by tannergooding** [comment]
Right, unintuitive C names, added a new entry that covers tau and there not being a corresponding `M_*` name in native.

**Comment by bartonjs** [comment]
## Unary Plus

Keep TResult, give up on the DIM

```C#
namespace System.Numerics;

// as-is
public interface IUnaryPlusOperators<TSelf, TResult>
    where TSelf : IUnaryPlusOperators<TSelf, TResult>
{
    static abstract TResult operator +(TSelf value);
}
```

## Comparison and Equality

Add TResult, lock TResult to bool at INumber.

```C#
namespace System.Numerics;

// add TResult
public interface IComparisonOperators<TSelf, TOther, TResult>
    : IComparable,
      IComparable<TOther>,
      IEqualityOperators<TSelf, TOther>
    where TSelf : IComparisonOperators<TSelf, TOther>
{
    static abstract TResult operator <(TSelf left, TOther right);
    static abstract TResult operator <=(TSelf left, TOther right);
    static abstract TResult operator >(TSelf left, TOther right);
    static abstract TResult operator >=(TSelf left, TOther right);
}

public interface IEqualityOperators<TSelf, TOther> : IEquatable<TOther>
    where TSelf : IEqualityOperators<TSelf, TOther>
{
    static abstract TResult operator ==(TSelf left, TOther right);
    static abstract TResult operator !=(TSelf left, TOther right);
}
```

## ITrigonometricFunctions

Add SinCosPi, move Atan2/Atan2Pi to the scalar-only interface

```C#
namespace System.Numerics;

public interface ITrigonometricFunctions<TSelf>
    where TSelf : ITrigonometricFunctions<TSelf>, IFloatingPoint<TSelf>
{
    static abstract TSelf Acos(TSelf x);
    static abstract TSelf AcosPi(TSelf x);
    static abstract TSelf Asin(TSelf x);
    static abstract TSelf AsinPi(TSelf x);
    static abstract TSelf Atan(TSelf x);
    static abstract TSelf AtanPi(TSelf x);
    static abstract TSelf Cos(TSelf x);
    static abstract TSelf CosPi(TSelf x);
    static abstract TSelf Sin(TSelf x);
    static abstract (TSelf Sin, TSelf Cos) SinCos(TSelf x);
    static abstract (TSelf SinPi, TSelf CosPi) SinCosPi(TSelf x);
    static abstract TSelf SinPi(TSelf x);
    static abstract TSelf Tan(TSelf x);
    static abstract TSelf TanPi(TSelf x);
}

partial interface IFloatingPointIeee754<TSelf>
{
    static abstract TSelf Atan2(TSelf y, TSelf x);
    static abstract TSelf Atan2Pi(TSelf y, TSelf x);
}
```

## IPowerFunctions and IRootFunctions

Match the C runtime with an "N" suffix when we're locking the second argument to int.

```C#
namespace System.Numerics;

public partial interface IRootFunctions<TSelf>
{
    static abstract TSelf RootN(TSelf x, int n);
}

```

## Math Constants

* Renamed IMathConstants to IFloatingPointConstants

```C#
namespace System.Numerics;

public interface IFloatingPointConstants<TSelf>
    where TSelf : IFloatingPointConstants<TSelf>, INumberBase<TSelf>
{
    public static TSelf E { get; }

    public static TSelf Pi { get; }

    public static TSelf Tau { get; }
}

partial interface IFloatingPoint<TSelf> : IFloatingPointConstants<TSelf> {}

// Decimal implements these explicitly
```

**Comment by tannergooding** [comment]
This was resolved in #71800
